src/rust/src/wallet.rs

```diff
@@ -135,33 +139,80 @@ impl Wallet {
         Wallet {
             key_store: KeyStore::empty(),
             wallet_tx_notes: HashMap::new(),
+            nullifiers: BTreeMap::new(),
             witness_tree: BridgeTree::new(MAX_CHECKPOINTS),
             last_observed: None,
             locked_notes: HashSet::new(),
             spent_notes: HashSet::new(),
         }
     }
 
-    pub fn checkpoint_witness_tree(&mut self) {
-        self.witness_tree.checkpoint();
+    pub fn checkpoint(&mut self, block_height: BlockHeight) -> bool {
+        if let Some(last_observed) = &self.last_observed {
+            if last_observed.height == block_height {
+                self.witness_tree.checkpoint();
+                true
+            } else {
+                false
+            }
+        } else {
+            // it's fine to checkpoint the empty tree.
+            self.witness_tree.checkpoint();
+            true
+        }
     }
 
-    pub fn rewind_witness_tree(&mut self) -> bool {
-        self.witness_tree.rewind()
+    pub fn rewind(&mut self) -> bool {
+        if let Some(last_observed) = self.last_observed.as_mut() {
+            // when we rewind, we need to remove the spentness marks
+            // from the nullifiers we're tracking
+            if self.witness_tree.rewind() {
+                last_observed.height = last_observed.height - 1;
+                last_observed.block_tx_idx = 0;
+                for (txid, n) in self.wallet_tx_notes.iter() {
+                    if n.tx_height > last_observed.height {
+                        self.spent_notes.retain(|o| o.txid != *txid)
```

Comment:
This isn't removing entries from the nullifier map, is that intentional?

Reply:
Yes; per my other comment, we need to preserve the `nullifier` data in order to determine conflicts. "Spent-ness" detection during note selection necessarily needs to check if any transaction the nullifier is present in, is itself present in a block in the main chain.

Reply:
Yes. There's a separate change that I want to make that will be part of #5568 which will remove txids for transactions that have expired from this list; it's more efficient to do that change in a PR there since locks will expire at the same time.